README.MD
# PDFGeneration Lib

## Rules

- Add your version/tag release information on the [change log](CHANGELOG.MD)
- If you require the Charts Type install `"vega": "5.17.0"` and `"sharp": "^0.26.3"`

## Getting started

- Add the library in your package.json under dependencies `"@thisisme/pdfgeneration-lib": "https://github.com/ThisIsMe-dev/PDFGeneration-lib#5.3.0"`
- npm install to install all your dependencies for your project

### Examples

- Create a pdf_req.js file in your service that exports the following:

#### Generates a PDF Report with Content

```JavaScript
generatePDFContent: (response, event, requestTimestamp, requestParams, dataSource) => {
console.log('Generating PDF Object');
const pdfLimiter = pdf.pdfLimiter;
const pdfHelpers = new pdf.PDFHelpers(event, requestTimestamp, dataSource, requestParams);
const reportHeaders = {
    businessInformation: {title: 'Business Information', new_page: true},
    linkedAddress: {title: 'Linked Addresses', new_page: true},
    aliases: {title: 'Known Aliases', new_page: false},
    notes: {title: 'Notes', new_page: true},
    linkedBusinesses: {title: 'Linked Businesses', new_page: false},
    linkedPersons: {title: 'Known Associates', new_page: true},
}; // These are custom headers for you service that the PDF needs to know are headers so it can adjust the look accordingly
const content = pdfHelpers.generateResultsFoundPDFStartingContent(reportHeaders, constants.PDFType.CONTENTS_OF_PAGE);
return populateRestOfReport(pdfLimiter, reportHeaders, pdfHelpers, response, content); // this is your custom function that handles the rest of your custom service's pdf
}
```

#### Generates a PDF Report without Content

```JavaScript
generateNoResultsPDFContent: (event, requestTimestamp, requestParams, dataSource) => {
const pdfHelpers = new pdf.PDFHelpers(event, requestTimestamp, dataSource, requestParams);
return pdfHelpers.generateNoResultsPDFContent();
}
```

#### Remember to include your modules

```JavaScript
const pdf = require('@thisisme/pdfgeneration-lib');
const constants = require('@thisisme/pdfgeneration-lib/app/constants');
```

## Current Types

### Document Types

```JavaScript
const PDFType = {
  DEFAULT: 0, // old style search/response
  CONTENTS_OF_PAGE: 1,
  NO_SERVICE_RESPONSE_HEADER: 2,
  COVER_AND_CONTENTS_OF_PAGE: 3,
};
```

### Document Line Types

```JavaScript
const PDFDocumentLineType = {
  DEFAULT_LINE: 0, // Just a placeholder, this currently does the same as a KEY_VALUE_LINE
  HEADER_LINE: 1, // A headerline (just a bit bolder than the normal lines)
  KEY_VALUE_LINE: 2, // The basic line we use for just a label and text line
  EMPTY_LINE: 3, // Functionally the same as END_LINE at the moment, this is just a more visible label for the user to understand what they are trying to accomplish
  ADDRESS_LINE: 4, // This splits a string on its commas into multiple lines
  COLUMN_INFO: 5, // This requires a value of object with rows defined in (SINGLE_COLUMN, DOUBLE_COLUMN, END_LINE) - this has a different type of header than META_INFO
  META_INFO: 6, // This requires a value of object with rows defined in (SINGLE_COLUMN, DOUBLE_COLUMN, END_LINE)
  SINGLE_COLUMN: 7, // This specifies that a row can only have 1 column
  DOUBLE_COLUMN: 8, // This specifies that you can have 2 columns
  END_LINE: 9, // Normally used to signal that your section ended
  GRID: 10, // Helps to create a list/table like display with columns, taking an array of objects
  TABLE_OF_CONTENTS_LINE: 11, // Line used to define table of contents page lines
  KEY_LINK_LINE: 12, // A line that allows the user to click to go to the provided URL
  INDICATIVE_BAR_LINE: 13, // Indicative bars designed with Credit Scores in mind
  IMAGE_LINE: 14, // A line that represents an image
  PAGE_BREAK: 15, // Skips to the top of a new page
}
```

### PDF Image Types

```JavaScript
const PDFImageType = {
  DEFAULT: 0, // Image(s) placed next to each other
  CENTER: 1, // Image positioned in the center of the line
}
```

CHANGELOG.MD
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2020-07-17

### POC

- This is the library in a POC state.

## [1.0.0] - 2020-07-25

### Initial Release

- This library is now being used by ThisIsMe's Company Credit Report v3 call.

## [2.0.0] - 2020-07-30

### The Clean Release

- Previous code base was unstable and heading down a spaghetti code path, a lot of refactoring was done to avoid unnecessary errors and confusion, this version also supports more line types.
- Library name also changed from `@ThisIsMe/pdfgeneration-lib` to `@thisisme/pdfgeneration-lib`.

## [3.0.0] - 2020-08-10

### The Fancy Release

- Text have been made smaller.
- Fancy headers have been brought in for defined Headers.
- Grid Type has been added.
- File Structure has been improved, not just using app.js but split logic into different files as welll to help improve readability and maintainability.
- Pipelines have been implemented to test code pushed up.

## [4.0.0] - 2020-08-29

### The start of document types

- Document types added.
- More line types added.
- Helper builders added.

## [4.0.1] - 2020-08-30

### Bug fix

- width size of rows on pdf reduced to prevent lines going over to next.

## [5.0.0] - 2020-10-08

### Image and Indicative Bar (Score)

- Some refactoring to ease reading and future maintainability
- New types

## [5.0.1] - 2020-10-11

### Bug fix

- Minor logging corrections

## [5.1.1] - 2020-10-16

### Minor change

- Report bucket setting changed env var

## [5.2.1] - 2020-10-16

### Minor change

- AWS_DEFAULT_REGION reference changed to TIM_AWS_DEFAULT_REGION

## [5.3.0] - 2020-11-04

### Cover Page & Bug fixes

- Code restructure
- Cover Page added as a PDF Type
- Fixed headers that aren't selected as new page headers
- Fixed async issue with older PDF Types
- Added functionality to build tables

## [5.3.1] - 2020-11-05

### Bug fix

- splitLongText function fixed, was not catering for long words

## [5.4.0] - 2020-11-08

### Image functionality improvements and Grid font selection

- images will be placed side by side when array of images
- custom font adjustments specifically for Grid types

## [5.5.0] - 2020-11-17

### Image functionality improvements

- Image types added (Default and Center)
- Center adds image in the center of the focused line
- Helper functions added for image creation
- Fixed some camel case occurrences

## [5.6.4] - 2021-02-21

### Bug fix

- Bug fix, "bwip-js": "2.0.12" introduces an `undefined` value bug, locking version to 2.0.11.

## [5.6.5] - 2021-03-02

## Allow front page with Search Params

- allow front page without search params | checks null state

## [5.6.6] - 2021-03-02

## Allow front page with Search Params - Fix for 5.64

- allow front page without search params | checks object length

## [5.7.0] - 2021-03-4

## Image descriptions

- y increments for images fixed, extending base pdfkit
- Single images now cater for descriptions
- Centered fixed
- Image code refactor

## [5.8.0] - 2021-04-14

## Pdf responds when upload complete

- The PDF location will now be received when upload to s3 was successful

## [5.9.0] - 2021-04-14

## QRCode generation fix | Unit tests added

- `Jest` added for unit tests
- `'use strict'` above all .js files
- QRCode generation function fixed, covered by unit test

## [5.10.0] - 2021-04-21

## imageDescriptions support for addImageLineFromBase64 

- added imageDescriptions support to addImageLineFromBase64 
- re-ordered the import modules to match external, internal flow

## [5.10.1] - 2021-06-19

## Removed QRCode generation

app/constants.js
'use strict';

const fs = require('fs');

const packageName = require('../package.json').name;

function getPackagePath() {
  let PACKAGE_PATH = `node_modules/${packageName}/`;
  if (!fs.existsSync(PACKAGE_PATH)) {
    PACKAGE_PATH = '';
  }
  return PACKAGE_PATH;
}

const PDFType = {
  DEFAULT: 0, // old style search/response
  CONTENTS_OF_PAGE: 1,
  NO_SERVICE_RESPONSE_HEADER: 2,
  COVER_AND_CONTENTS_OF_PAGE: 3,
};

const PDFDocumentLineType = {
  DEFAULT_LINE: 0,
  HEADER_LINE: 1,
  KEY_VALUE_LINE: 2,
  EMPTY_LINE: 3,
  ADDRESS_LINE: 4,
  COLUMN_INFO: 5,
  META_INFO: 6,
  SINGLE_COLUMN: 7,
  DOUBLE_COLUMN: 8,
  END_LINE: 9,
  GRID: 10,
  TABLE_OF_CONTENTS_LINE: 11,
  KEY_LINK_LINE: 12,
  INDICATIVE_BAR_LINE: 13,
  IMAGE_LINE: 14,
  PAGE_BREAK: 15,
  CHART_LINE: 16,
};

const PDFTableType = {
  DEFAULT: 0,
  COVER: 1,
};

const PDFTableColumnTextAlign = {
  LEFT: 0,
  RIGHT: 1,
};

const PDFImageType = {
  DEFAULT: 0,
  CENTER: 1,
};

module.exports = Object.freeze({
  PACKAGE_PATH: getPackagePath(),
  // Where the page starts
  TOP_OF_PAGE_Y: 80,
  X_START: 20,

  // Normal Information
  INCREMENT_MAIN_Y: 18,

  // Usually meta data of sections
  INCREMENT_SUB_Y: 14,

  // Spacing between text and the line below it
  INCREMENT_UNDERLINE: 15,

  HEADER_FONT_SIZE: 16,
  NORMAL_FONT_SIZE: 10,

  PDFType,
  PDFDocumentLineType,
  PDFTableType,
  PDFTableColumnTextAlign,
  PDFImageType,

  // Colors used in Footer and Disclaimer are not currently covered here
  PDFColors: {
    NORMAL_COLOR: '#888888',
    INDICATIVE_COLOR: 'black',
    TEXT_IN_NORMAL_COLOR: 'white',
  },

  PDFDocumentLineRules: { // Future Idea
    ALWAYS_NEW_PAGE: 0,
  },

  PDF_TEXT: {
    REPORT_AUTHOR: 'ThisIsMe (Pty) Ltd',
    REPORT_HEADERS: ['Search Parameters:', 'Service Response:'],
  },
});
